helmexec/helmexec.go

```diff
@@ -11,4 +13,11 @@ type Interface interface {
 	DeleteRelease(name string) error
 
 	DecryptSecret(name string) (string, error)
+
+	// unit testing
+	exec(args ...string) ([]byte, error)
+	setRunner(runner Runner)
+	getExtra() []string
+	getKubeContent() string
+	getWriter() io.Writer
```

Comment:
You've made me realize a good point!

TL;DR; For this case, `func NewHelmExec(writer io.Writer, kubeContext string) Interface {` should return the concrete type, rather than the interface so that it looks `func NewHelmExec(writer io.Writer, kubeContext string) execer {`.

Doing so would probably make `getExtra`, `setRunner`, `getKubeContent`, `getWriter` unnecessary, because then you can just reference the fields directly in tests you've added.

`type Interface` should be declared in the `main` package hence `./helmexec/helmexec.go` should be moved to `./helmexec.go`.

I can't say this pattern should be enforced everywhere, but you can consult [this reddit and the article](https://www.reddit.com/r/golang/comments/5cos3x/accept_interfaces_return_structs/) for other pros of doing it, too.

Reply:
I was thinking the same thing about the return time, changed. However, I am not sure I follow how the `Interface` should now go into the main package. The `state` package needs access to it and it seems to also be more at home in the `helmexec` package since it describes its functionality.